src/oc_erchef/apps/chef_objects/src/chef_cbv_cache.erl

```diff
@@ -0,0 +1,228 @@
+                tid = ets:new(chef_cbv_cache, [set, private, {read_concurrency, true}])
+               }}.
+handle_call(stop, _From, State) ->
+    exit(whereis(cbv_cache_breaker), kill),
+    {stop, normal, ok, State};
+handle_call(_Msg, _From, #state{enabled = false} = State) ->
+    {reply, undefined, State};
+handle_call({get, Key}, _From, #state{tid = Tid, claims = Claims} = State) ->
+    case dict:take(Key, Claims) of
+        error ->
+            % Nobody is currently working on it, so it's either here or not.
+            case ets:lookup(Tid, Key) of
+                [{_, Value}] -> {reply, Value, State};
+                [] -> {reply, undefined, State}
+            end;
+        {Pid, Claims1} ->
+            case is_process_alive(Pid) of
+                true -> {reply, retry, State};
+                false ->
+                    case ets:lookup(Tid, Key) of
+                        [{_, Value}] -> {reply, Value, State};
+                        [] -> {reply, undefined, State#state{claims = Claims1}}
+                    end
+            end
+    end;
+handle_call({claim, Key}, {From, _Tag}, #state{claims = Claims} = State) ->
+    case dict:take(Key, Claims) of
+        error -> % Nobody has claimed it yet
+            {reply, ok, State#state{claims = dict:store(Key, From, Claims)}};
+        {From, _Claims1} ->
+            % This caller has already claimed it, that's fine..
+            {reply, ok, State};
+        {_OtherPid, _Claims1} -> % Someone has already claimed it, so caller should retry the GET
+            {reply, retry, State}
+    end;
+handle_call({put, Key, Value}, { From, _Tag }, #state{tid = Tid, ttl = TTL, claims = Claims} = State) ->
+    case dict:take(Key, Claims) of
+        error ->
+            % Mis-usage. Claim must be called before put.
+            {reply, {error, no_claim}, State};
+        { From, Claims1 } ->
+            % It is claimed and by this caller, they can go ahead and put.
+            % Update claims dictionary in state to remove the claim on this key
+            insert_into_cache(Tid, Key, Value, TTL),
+            {reply, ok, State#state{claims = Claims1}};
+        { _OtherPid, _Claims1 } ->
+            % Mis-usage: Someon else claimed it and is still active/working on it,
+            % but we have been invoked anyway. `put` should only be invoked when the caller
+            % has first invoked `claim`.
+            {reply, {error, already_claimed}, State }
+    end.
+
+insert_into_cache(Tid, Key, Value, TTL) ->
+    case ets:insert_new(Tid, {Key, Value}) of
+        true ->
+            % - we see that often the pattern is that many requests for new keys
+            % come in clusters when traffic is first directed to the server.
+            % We can't avoid the initial cluster and corresponding CPU spike
+            % when we have nothing cached, but staggering the expirations after that
+            % will reduce CPU/VM utilization spikes that can impact the overall system.
```

Comment:
considering removing this behavior - now that we have synchronization to prevent a flood of callers from doing the same work, it may no longer be necessary. (On the other hand, it probalby doesn't hurt...)

Reply:
Testing shows that this is still helpful under load, and does allow the server fail point to be at a higher rate of traffic.